File: src/ru/hudspeed/pro/HeadingSelection.java
package ru.gpsantiradar.app;

/** Keeps the immediate visual course independent from alert-direction smoothing. */
public final class HeadingSelection {
    public final float visualHeading;
    public final float alertHeading;

    private HeadingSelection(float visualHeading, float alertHeading) {
        this.visualHeading = visualHeading;
        this.alertHeading = alertHeading;
    }

    public static HeadingSelection forStrelka(float gpsHeading, float radarHeading,
                                              float proposedRadarHeading,
                                              boolean scanRequested,
                                              boolean hasPreviousRadarFix) {
        float alertHeading = gpsHeading;
        if (scanRequested && hasPreviousRadarFix) {
            alertHeading = proposedRadarHeading;
        } else if (!Float.isNaN(radarHeading)) {
            alertHeading = radarHeading;
        }
        return new HeadingSelection(gpsHeading, alertHeading);
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/HeadingSelection.kt
package ru.gpsantiradar.app

/** Keeps the immediate visual course independent from alert-direction smoothing. */
class HeadingSelection private constructor(
    val visualHeading: Float,
    val alertHeading: Float
) {
    companion object {
        fun forStrelka(
            gpsHeading: Float, radarHeading: Float, proposedRadarHeading: Float,
            scanRequested: Boolean, hasPreviousRadarFix: Boolean
        ): HeadingSelection = HeadingSelection(
            gpsHeading,
            when {
                scanRequested && hasPreviousRadarFix -> proposedRadarHeading
                !radarHeading.isNaN() -> radarHeading
                else -> gpsHeading
            }
        )
    }
}
